Storage package

tomponline opened the pull request:
Based on https://github.com/lxc/lxd/pull/6274

Includes https://github.com/lxc/lxd/pull/6317
Includes https://github.com/lxc/lxd/pull/6360

tomponline commented:
@stgraber as discussed, this is my PR of your storage branch where I will work.

tomponline commented:
@stgraber I've added volume create and copy support for dir backend, please can you take a look and check the general direction. Tests should pass OK.

tomponline commented on lxd/storage/drivers/driver_dir.go:
```diff
@@ -0,0 +1,209 @@
+		srcVolPath = GetStoragePoolVolumeSnapshotMountPoint(srcPool, srcVolName)
+	} else {
+		srcVolPath = GetStoragePoolVolumeMountPoint(srcPool, srcVolName)
+	}
```
@stgraber I called this function `CopyVolume` rather than `CopyCustomVolume` as wanted to keep it generic as possible in the driver implementation. However the helper functions used in the previous implementation; `GetStoragePoolVolumeSnapshotMountPoint` and `GetStoragePoolVolumeMountPoint` assume that the volume path includes "custom" and further down `GetStoragePoolVolumeSnapshotMountPoint` assumes "custom-snapshots". 

Is it worth making more generic versions of these helper functions that would accept the `VolumeType` as an argument and build the path based on that, or do you think that copying an "instance" volume would be sufficiently different as to need a separate "CopyInstanceVolume" function?

stgraber replied:
See the comment I left on an earlier commit about `CreateVolumeFromCopy` and the workflow I envision for containers/virtual-machines creation from images through it.

stgraber replied:
Ah and worth noting that VolumeType does differentiate between volumes and snapshots, so we shouldn't need IsSnapshot at all.

tomponline replied:
@stgraber the constants you've added so far for VolumeType do not included the snapshot variants, was it your intention to add more for those variants, such as `VolumeTypeCustomSnapshot`?

tomponline replied:
@stgraber I've solved this with the `GetVolumeMountPoint` function that detects the "/" in the volume name and treats it as a snapshot, so no new constants needed.

lxc-jenkins commented:
Testsuite passed

stgraber commented on lxd/storage/drivers/interface.go:
```diff
@@ -0,0 +1,52 @@
+	GetResources() (*api.ResourcesStoragePool, error)
+
+	// Volumes
+	CreateVolume(volType VolumeType, volID int64, volName string, op *operations.Operation) error
```
Hmm, so based on past discussions, I think our best bet is to change `CreateVolume` and `CopyVolume` to be:
```go
CreateVolume(volType VolumeType, contentType contentType, name string, filler func(path string) error, op *operations.Operation) error
CreateVolumeFromCopy(volType VolumeType, name string, srcType VolumeType, srcName string, snapshots bool, op *operations.Operations) error
```

The filler function, if set, is called with the path to either a mounted directory (if contentType is filesystem) or a mapped block device (if contentType is block).

CreateFromCopy lines up with the naming on the Pool interface and takes enough details to create a volume as a copy of another one or from a snapshot of another one. It would also allow creating a new container or vm volume from an image volume.

tomponline replied:
@stgraber ah yes, right, although did we say that we could detect the contentType from the path by detecting if it is a block device and if not assuming it is a mounted path (could check that it is a directory too).

As for CreateVolumeFromCopy, if depending on the outcome of the discussions above regarding VolumeTypes for snapshots, we would not need the snapshot boolean here. Then we would just call this function multiple times from lxdBackend based on how many snapshots (and the original volume itself) there are.

tomponline replied:
@stgraber I've found that we can actually use the volume name itself to indicate if its a snapshot or not, so no need for additional types or arguments.

stgraber replied:
You're going to need `srcType` in the case where you're creating a new volume of type container or virtual-machine from a volume of type image.

stgraber replied:
Our discussion on Monday was about the arguments passed to `filler` which indeed do not need the contentType passed through to it.

We do need `contentType` for Create itself, as otherwise you have no way to indicate whether you're creating an image volume for a container (filesystem) or a VM (block).

tomponline replied:
@stgraber I see, is contentType only needed because the "image" volume type doesn't indicate whether it is a VM image or a container image? Is it worth perhaps instead adding a concept of a VM image and then pass this via the VolumeType arg. As my understanding of this now is that we could infer the volume type if block for VM VolumeTypes and FS for containers, but we cannot at this time infer block/FS for the Image VolumeType.

tomponline replied:
@stgraber another question, is it possible to do cross-pool copies, ie copy a volume on one pool to another pool of a different type? If so, in this case, when should the source volume be mounted (assuming it is of type that requires a per-volume mount, such as LVM and ZFS)? I've been looking in the source now and I cannot find the logic that handles mounting source volumes.

tomponline replied:
@stgraber as discussed, my understanding is that whilst cross-pool copies are possible it should not be done via CreateVolumeFromCopy, but instead using the `MigrateCustomVolume` and `CreateCustomVolumeFromMigration` functions.

stgraber commented on lxd/storage/drivers/interface.go:
```diff
@@ -0,0 +1,52 @@
+
+	// Migration
+	MigrationType() migration.MigrationFSType
+	PreservesInodes() bool
```
I think we need an extra `OptimizedImages() bool` symbol as well as a `HasVolume(volType VolumeType, name string) bool` one.

This will then allow a workflow like this from the lxdBackend side during a container creation from an image:
 - OptimizedImages()
   - If true
     - Call HasVolume to see if we have an image volume, note that for ceph & zfs, we'll need that to return `true` for deleted volumes that we kept around for refcounting reasons (`deleted/images/XYZ`).
     - If we don't have an existing volume, call `CreateVolume` for `VolumeTypeImage` and set the `filler` so we can create that missing image volume
     - Then call `CreateVolumeFromCopy` with `srcType` of `VolumeTypeImage` and `srcName` set to the image fingerprint
   - If false
     - Call CreateVolume with a filler set to copy or unpack the image depending on whether filesystem or block

The result is a container created from an image, whether the filesystem has optimized support or not, regardless of whether we're dealing with block or filesystem and dealing with the deleted image trick we have for zfs/ceph and with minimal logic in the Driver and centralized image creation logic in the Backend.

tomponline replied:
@stgraber can you remind me what `OptimizedImages` means, is it when a storage driver caches an unpacked image tarball in a volume with a snapshot so that when creating an instance one can just create a volume from the cached snapshot?

Also, can you explain further about ceph and zfs and the ref counting you referred to? Thanks

stgraber replied:
Correct, OptimizedImages would return true if we're using an image type volume that gets cloned. So this would be true for zfs, lvm, btrfs and ceph, false for dir and cephfs.

Both ZFS and CEPH do not allow the destruction of a volume which is used as the parent of another volume. In both cases, rather than fail the deletion of the image or container because of that refcount, we instead rename the image or container to a name indicating it's been deleted. Then for images, if we're later asked to import that same image again, we can just resurrect the deleted dataset rather than unpacking it again.

tomponline replied:
@stgraber thanks, I've added those new functions.

stgraber commented on lxd/storage/backend_lxd.go:
```diff
@@ -0,0 +1,353 @@
+		return err
+	}
+
+	revert := true
```
Should be called `revertDatabase`. I've recently gone with an approach of having one of those revert booleans per reversible action so it's easier to understand what's going on and avoids some of the issues we've had with sub-reverts in zfs.

tomponline replied:
@stgraber ok will do. Out of interest, is there a reason we dont use transactions on the database and wait until the on disk stuff has happened before committing the transaction for the whole operation. Does this cause locking issues with sqlite's table level locking, or is it that it would require passing around the same db handle so different functions can read their own writes inside the transaction (to get IDs etc).

tomponline replied:
OK will leave it out.

stgraber commented on lxd/storage/backend_lxd.go:
```diff
@@ -0,0 +1,353 @@
+	}
+
+	// Get the source pool's ID for use later copying snapshots.
+	srcPoolID, err := b.state.Cluster.StoragePoolGetID(vol.Source.Pool)
```
Feels like we could probably stuff that in the struct at loading/creation time, would save us from having to query it, potentially multiple times down the line.

tomponline replied:
@stgraber yes agreed, I'd also like to look into whether we can get the DB row create function to return the auto inc ID rather than having to query for it again.

tomponline replied:
@stgraber this has been added.

stgraber commented on lxd/storage/backend_lxd.go:
```diff
@@ -0,0 +1,353 @@
+	}
+
+	// Create database entry for new storage volume.
+	err = VolumeDBCreate(b.state, b.name, vol.Name, vol.Description, vol.Type, false, volConfig)
```
Missing revert logic in this function.

tomponline replied:
I've added revert logic now.

stgraber commented on lxd/storage/backend_lxd.go:
```diff
@@ -0,0 +1,353 @@
+	}
+
+	// Check the supplied config and remove any fields not relevant for destination pool type.
+	volConfig, err := VolumePropertiesTranslate(vol.Config, b.driver.Name())
```
So this we'll need to do differently, I don't want to have the supported pool config and volume config be stored outside of the driver. Adding a support pool or volume option should only require a change to the driver itself and to nothing else.

tomponline replied:
@stgraber I approached this by adding a `removeUnknownKeys` argument to the new volume validation functions, when doing a copy of a volume the supplied config is validated with `removeUnknownKeys` set to true such that keys not recognised by the destination pool's driver are removed.

stgraber commented on lxd/storage/backend_lxd.go:
```diff
@@ -0,0 +1,353 @@
+
+	// If we are copying snapshots, create database entry for each snapshot.
+	var copySnapshots []string
+	if !vol.Source.VolumeOnly {
```
This is doing 1+2*N full DB read transactions (where N is the number of snapshots), this seems very heavy and something we could probably optimize so we can get a slice of structs that include the bits we need in one shot.

tomponline replied:
@stgraber yes this isn't particularly efficient, although it is somewhat better than what is currently there, as each iteration calls `storagePoolVolumeSnapshotCopyInternal` which then does a query to `StoragePoolGetID` which is avoided here. I wasn't sure how far you wanted this rework to go, in terms of whether it was acceptable to start changing queries in the cluster package too.

tomponline replied:
@stgraber this has been updated to be more efficient now.

stgraber commented on lxd/storage/drivers/driver_dir.go:
```diff
@@ -0,0 +1,209 @@
+		return err
+	}
+
+	err = d.initQuota(volPath, volID)
```
Forgot about that part, the fact that project quotas require an integer be associated with the quota is annoying, I really would have liked to not expose such low level in there, especially as it will be meaningless in 90% of the drivers...

What would you think of passing a function pointer to a `GetVolumeID(volType VolumeType, name string) (int64, error)` function at Driver instantiation time? This would then allow the storage driver to call `d.getVolumeID(volType, name)` whenever a unique identifier for the volume is needed rather than having us have to pass it around to every function which may eventually need it.

tomponline replied:
@stgraber sounds good to me :)

tomponline replied:
@stgraber this has been added.

stgraber commented on lxd/storage/utils.go:
```diff
@@ -360,3 +384,253 @@ func GetStorageResource(path string) (*api.ResourcesStoragePool, error) {
+var SupportedPoolTypes = []string{"btrfs", "ceph", "cephfs", "dir", "lvm", "zfs"}
+
+// StorageVolumeConfigKeys config validation for btrfs, ceph, cephfs, dir, lvm, zfs types.
+var StorageVolumeConfigKeys = map[string]func(value string) ([]string, error){
```
We're going to want that split between generic config options and driver specific config options.

The drivers will be responsible for validating their own config and no mention of those config keys should be present outside of the driver.

The backend will be responsible for handling the generic config options (such as `security.shifted` and `security.unmapped`) and we'll filter those out from the data exposed to the drivers so we don't need them all be aware of config options that don't concern them.

tomponline replied:
@stgraber ok, so are you thinking that there would be a non-exported function attached to the lxdBackend called `volumePropertiesTranslate` which would then validate generic parts of the config, and inside that function call the storage pool's driver function `VolumePropertiesTranslate` function passing it a modifier config with the generic fields removed for the driver to validate/remove and then the return from that would be merged back into the generic fields and returned to the caller?

stgraber replied:
Yeah, in my head, we'd just have:
 - A map in the `storage` package for the generic keys
 - A private `validateVolume` options in `lxdBackend` which ignore additional keys and chains to the public function from the driver
 - A public `ValidateVolume` function in the Driver which takes a `StorageVolumesPost` with its config filtered so the keys in the map are filtered out.

tomponline replied:
@stgraber the approach I've gone with is similar, but more like the device validation.

I've added a `validateVolumeCommonRules` and `validateVolume` function to the common driver. The former generates the common validation rules, and the latter is called from each driver's exported `ValidateVolume` function. It will take the `validateVolumeCommonRules` generated rules as a base and then merge any custom rules supplied by the driver.

It will then run through all the validation rules inside the driver (both common and driver specific). 

This achieves the same effect, but just defines the config in the `common` driver rather than in `lxdBackend`.

tomponline replied:
@stgraber I'm not clear as to whether the `size` property is valid for `dir` volumes. The defaults seem to populate it with an empty string, and the validation function doesnt include dir type when validating size. But project quotas suggest it should have a size property. Can you clarify?

tomponline replied:
@stgraber confirmed on IRC that the size property is valid for dir, have added support for that now, and made the "size" property a common config field for all storage drivers.

tomponline commented on lxd/storage/drivers/utils_paths.go:
```diff
@@ -0,0 +1,42 @@
+package drivers
```
@stgraber sounds good, that was my thinking also. However there is not yet constants for snapshots, so either we would need some new constants, such as "VolumeTypeCustomSnapshot" or we would need an extra flag or function to indicate a snapshot of a certain volume type.

tomponline replied:
@stgraber GetVolumeMountPoint has been added without the need for extra snapshot variant constants.

tomponline commented on lxd/storage/drivers/interface.go:
```diff
@@ -0,0 +1,52 @@
+type VolumeType string
+
+// VolumeTypeCustom represents a custom storage volume.
+const VolumeTypeCustom = VolumeType("custom")
```
@stgraber there are similar typea already defined in the db package, such as `StoragePoolVolumeTypeNameContainer` that is "container". There is a conversion function that converts from these string types to the DB int types, however as we are now using `VolumeTypeContainer` type this cannot be used with the existing conversion function (which I need to lookup the volume ID).

I can create another conversion function that takes `VolumeType` and converts to DB int, or perhaps we could use the existing db string constants instead, and then we can just append an "s" to the parts where its needed for the storage paths. What do you think?

stgraber replied:
Except for the part where the plural of `custom` is still `custom` in our case :)

I think we may need a `volumetype` package which gives us a struct for all of:
 - custom
 - custom snapshot
 - container
 - container snapshot
 - image
 - virtual machine
 - virtual machine snapshot

With the DB type for each of them, whether it's a snapshot or not, the API name, on-disk name, ... And port everything to using that.

tomponline replied:
@stgraber OK I'll keep that under advisement, for now I've avoided the need for another conversion function because after I found out that there was not a need to support copying different types of volumes to one and other, I was then able to hard code the new volume types based on the context of the function, `CreateCustomVolumeFromCopy` for example.

tomponline commented on lxd/storage/drivers/utils.go:
```diff
@@ -139,3 +141,20 @@ func vfsResources(path string) (*api.ResourcesStoragePool, error) {
+// GetVolumeMountPoint returns the mount path for a specific volume based on its pool and type and
+// whether it is a snapshot or not.
+// For VolumeTypeImage the volName is the image fingerprint.
+func GetVolumeMountPoint(poolName string, volType VolumeType, volName string) string {
```
@stgraber here is the unified GetVolumeMountPoint, with its detection of snapshots via volume name. I added some tests below too.

tomponline commented on lxd/storage/backend_lxd.go:
```diff
@@ -0,0 +1,351 @@
+
+	// tomp TODO is this needed?
+	if ourMount {
+		defer b.driver.Unmount()
```
@stgraber if the expectation is that all storage pools will be mounted at start or create time, then is this part still needed?

stgraber replied:
Yes, the expectation is that both `GetByName` and `Create` will return you a `Pool` but neither will mount it for you. LXD on startup will call `GetByName` on all pools, followed by a `Mount()` call of each of them. Similarly, when calling `Create()`, you should follow that with `Mount()` if you intend to use it.

tomponline commented on lxd/storage/backend_lxd.go:
```diff
@@ -0,0 +1,418 @@
+}
+
+// DeleteImage removes an image from the database and underlying storage device if needed.
+func (b *lxdBackend) DeleteImage(fingerprint string, op *operations.Operation) error {
```
@stgraber I've added DeleteImage now, although in the current dir implementation I cant see that it actually does anything except remove the DB entry. Is that expected because it doesn't need to create an image volume?

stgraber replied:
There shouldn't even be a DB entry in the dir case.

Anything which has OptimizedImages() return false shouldn't have image type volumes at all.

tomponline replied:
@stgraber I see thanks. I guess there is no harm in trying to remove it though. But that does explain why in my tests on the current implementation I couldnt see that row being created.

stgraber commented on lxd/storage/load.go:
```diff
@@ -0,0 +1,120 @@
+	}
+
+	// Define a function to retrieve a volume ID for a volume Name for use in driver.
+	volIDFunc := func(project string, volType drivers.VolumeType, volName string) (int64, error) {
```
The drivers shouldn't have the concept of projects at all, so if that function is to be exposed there, that's going to be a problem.

tomponline replied:
@stgraber the concept of projects appears to be hardcoded as "default" in the current storage layer. I can certainly hardcode "default" into the volIDFunc itself so that its not needed as a param, however it would be good to understand how projects will work with the storage layer in the future, as the volIDFunc is defined when the storage pool is instantiated and so would work fine if the storage pool was linked to a project, but wouldn't work if the project was used on a per-volume basis.

tomponline replied:
@stgraber as discussed, the volName will contain the project name in the future separated by an underscore.

lxc-jenkins commented:
Testsuite passed

lxc-jenkins commented:
Testsuite passed

stgraber commented on lxd/storage/drivers/interface.go:
```diff
@@ -0,0 +1,65 @@
+
+	// Volumes.
+	CreateVolume(volType VolumeType, contentType ContentType, volName string, filler func(path string) error, op *operations.Operation) error
+	CreateVolumeFromCopy(volType VolumeType, volName string, srcVolType VolumeType, srcVolName string, op *operations.Operation) error
```
Both `CreateVolume` and `CreateVolumeFromCopy` should actually be passed the Config map from api.StorageVolumesPost as there are driver specific information in there.

I just thought of it now thinking of how we'll be creating a "fs" contentType on LVM/CEPH where this involves the driver creating a block device of a configured size and creating the filesystem on it. Both size and filesystem are configurable at the pool level OR at the volume level, so the level config must be made available to the Create functions.

tomponline replied:
Add `volConfig map[string]string` to those two functions and passed vol.Config through from lxdBackend.

stgraber commented on lxd/storage/load.go:
```diff
@@ -0,0 +1,122 @@
+	}
+
+	// Define a function to retrieve a volume ID for a volume Name for use in driver.
+	volIDFunc := func(volType drivers.VolumeType, volName string) (int64, error) {
```
Given you have two near identical copies of this, maybe make a function which takes in the poolID and return you this function, then use it in both CreatePool and GetPoolByName?

tomponline replied:
Good idea, have implemented.

stgraber commented on lxd/storage/backend_lxd.go:
```diff
@@ -0,0 +1,429 @@
+		return nil
+	}
+
+	err = b.state.Cluster.StoragePoolVolumeDelete(project, fingerprint, db.StoragePoolVolumeTypeImage, b.id)
```
Given that `project` is only ever passed to StoragePoolVolumeDelete now, I would just pass `"default"` directly to the function.

tomponline replied:
Will do.

stgraber commented on lxd/storage/drivers/driver_dir.go:
```diff
@@ -0,0 +1,293 @@
+	// If the volume being removed is a snapshot, check if it is the last one being removed for
+	// the volume parent and if so then remove the parent volume's snaphots directory.
+	if shared.IsSnapshot(volName) {
+		// Extract just the snapshot name from the volume name and then remove that suffix
```
Looks to me like this entire logic isn't `dir` specific, it's effectively just about removing a directory if it's now empty, should be its own helper function so other drivers can use it.

tomponline replied:
@stgraber fair point, will do.

stgraber commented on lxd/storage/drivers/driver_dir.go:
```diff
@@ -0,0 +1,293 @@
+
+	// If the volume being removed is a snapshot, check if it is the last one being removed for
+	// the volume parent and if so then remove the parent volume's snaphots directory.
+	if shared.IsSnapshot(volName) {
```
I don't like the idea of us mixing normal volumes and snapshots together. It's something we've done everywhere in the past and that's come to bite us since. Things are also going to get weird with zfs, lvm and ceph where there is a strong distinction between volumes and snapshots.

Also be careful about quotas on dir backend, we want all snapshots to share the quota group of their volume, so we need to pass the volume name, not the snapshot name to getVoldID.

I think it'd be best if we had separate functions to create/delete/rename/restore snapshots.

We'll also want to make sure that deleting a volume which still has snapshots fails. This should never happen as the backend should do the right thing and delete the snapshots first, but we've seen weird inconsistencies before.

stgraber replied:
Thinking about this again, it may actually be fine to use the same functions as we have distinct volume types now.

We however shouldn't be using shared.IsSnapshot as we can and should distinguish based on voltype.

tomponline replied:
@stgraber there isn't a vol type for snapshot, so I'm not clear how these would help indicate a snapshot?

stgraber replied:
Okay, so clearly my memory was failing me, I thought we had distinct volume types for regular vs snapshots, we don't, so we obviously can't rely on them. So I think that convinced me to go back to separate functions then, even if they end up internally sharing a lot of logic through shared private functions.

CreateVolumeSnapshot (take a snapshot)
DeleteVolumeSnapshot (delete a snapshot)
RenameVolumeSnapshot (rename a snapshot)

tomponline replied:
@stgraber OK, I have added those functions to the interface and modified this function to not only delete the actual parent volume (and not the snapshots) as well as checking all the snapshots have been removed first.

stgraber commented on lxd/storage/drivers/driver_dir.go:
```diff
@@ -0,0 +1,293 @@
+}
+
+// CreateVolumeFromCopy provides same-pool volume copying functionality.
+func (d *dir) CreateVolumeFromCopy(volType VolumeType, volName string, srcVolType VolumeType, srcVolName string, op *operations.Operation) error {
```
I think we'll need an extra `snapshots bool` to control whether we copy snapshots or not.

tomponline replied:
@stgraber this function is only called to copy a single low level volume (either the parent or a single snapshot), so I'm not clear what the snapshots boolean would do?

stgraber replied:
Most drivers will not let you copy individual snapshots.

So you effectively have the following options:
 - Copy a volume and all its snapshots
 - Copy a volume and none of its snapshots
 - Copy a volume snapshot as a new volume

You can't copy a volume with no snapshot and then decide to copy the snapshots later.

Copying subset of snapshots is also unlikely to be consistently supported and isn't something we currently allow in the API, so a snapshot slice isn't really helpful.

stgraber replied:
As for the copy from snapshot case. Since we want to be explicit about snapshot use, I think we'll want a `srcVolSnapshot string` argument. If it's empty, we copy a whole volume with or without its snapshots (depending on `snapshots bool`). If it's not empty, then we're copying from a volume snapshot into a new volume (`snapshots bool` is meaningless in that case).

tomponline replied:
@stgraber said in IRC:

"

[the drivers] all of them have some way of listing snapshots
--
21:30:50 | <stgraber> | tomponline: we'll need to skip internal snapshots (those we create during copy and migration for zfs and ceph) but we had that kind of logic before too

for dir/btrfs/cephfs, it's just a directory lookup, for zfs/ceph/lvm, the internal commands let you  get the list
"

tomponline replied:
@stgraber I've updated dir.CreateVolumeFromCopy and lxdBackend.CreateCustomVolumeFromCopy as we discussed. The only change I made was to change the `srcVolSnapshot` arg to `srcIsSnapshot` boolean instead. This seemed to achieve both the explicit notification to the driver that the source was a snapshot without the need to pass the same volume name (i.e volume/snapshot) in srcVolName and srcVolSnapshot.

I've used the srcIsSnapshot argument to detect when to not try and copy snapshots, even if copySnapshots is true.

stgraber commented on lxd/storage/utils.go:
```diff
@@ -360,3 +384,253 @@ func GetStorageResource(path string) (*api.ResourcesStoragePool, error) {
 }
+
+// VolumeTypeNameToType converts a volume type string to internal code.
+func VolumeTypeNameToType(volumeTypeName string) (int, error) {
```
This is going to fail if passed any of the snapshot types. Should it return a bool to indicate whether it's a snapshot or not?

stgraber replied:
Ignore me :)

stgraber commented on lxd/storage/utils.go:
```diff
@@ -360,3 +384,253 @@ func GetStorageResource(path string) (*api.ResourcesStoragePool, error) {
+	// already exist.
+	volumeID, _ := s.Cluster.StoragePoolNodeVolumeGetTypeID(volumeName, volumeType, poolID)
+	if volumeID > 0 {
+		return fmt.Errorf("a storage volume of type %s does already exist", volumeTypeName)
```
Error message nit but should be `A storage volume of type %s already exists`

stgraber commented on lxd/storage/utils.go:
```diff
@@ -360,3 +384,253 @@ func GetStorageResource(path string) (*api.ResourcesStoragePool, error) {
+}
+
+// SupportedPoolTypes the types of pools supported.
+var SupportedPoolTypes = []string{"btrfs", "ceph", "cephfs", "dir", "lvm", "zfs"}
```
We'll want to make sure we don't rely on this for any of the new code as we won't want such a list in the main `storage` package. Only `drivers` should tell us what backend exists.

stgraber replied:
We'll actually want some kind of `SupportedDrivers()` function the drivers package which should give us:
 - Name of the driver
 - Version
 - If it's currently usable (tools/kernel support present)
 - What can it store (list of volume types)
 - Whether it's a local storage or remote

This is all information we need for `lxd init` so we can offer the right set of storage drivers to the user at setup time. We may extend that description struct with some more bits based on what's needed to driver the `lxd init` questions.

stgraber commented on lxd/storage_volumes.go:
```diff
@@ -1040,43 +1040,65 @@ func storagePoolVolumeTypeDelete(d *Daemon, r *http.Request, volumeTypeName stri
+		case storagePoolVolumeTypeImage:
+			err = pool.DeleteImage(volumeName, nil)
+		default:
+			return response.BadRequest(fmt.Errorf(`Storage volumes of type "%s" `+
```
I tend to prefer we don't wrap such error strings, makes it much easier to grep.

tomponline replied:
Have fixed this.

stgraber commented on lxd/storage_volumes.go:
```diff
@@ -1040,43 +1040,65 @@ func storagePoolVolumeTypeDelete(d *Daemon, r *http.Request, volumeTypeName stri
+		case storagePoolVolumeTypeImage:
+			err = s.ImageDelete(volumeName)
+		default:
+			return response.BadRequest(fmt.Errorf(`Storage volumes of type "%s" `+
```
Same about wrapping

tomponline replied:
Fixed.

stgraber commented:
Ok, so for migration, the state of things as far as what's part of the requests is effectively:

## Migration flow
Request to source for instance migration (same for instance snapshot minus Live):
 - Migration (=true)
 - Live
 - InstanceOnly
 - Target (implies "push" mode)
    - Certificate
    - Operation
    - Websockets

Request to source for volume migration:
 - Migration (=true)
 - Target
    - Certificate
    - Operation
    - Websockets
 - VolumeOnly

Request to target for instance migration (same for instance snapshot):
 - Type (=migration)
 - Certificate
 - Mode
 - Operation
 - Websockets
 - InstanceOnly
 - Refresh

Request to target for volume migration:
 - Type (=migration)
 - Certificate
 - Mode
 - Operation
 - Websockets
 - VolumeOnly

Data exchanged over migration websocket for instances:
 - fs
 - criu
 - idmap
 - snapshotNames
 - snapshots
    - name
    - localConfig
    - profiles
    - ephemeral
    - localDevices
    - architecture
    - stateful
    - creation_date
    - last_used_date
 - predump
 - rsyncFeatures
 - refresh
 - zfsFeatures

Data exchanged over migration websocket for volumes:
 - fs
 - idmap
 - snapshotNames
 - snapshots
    - name
    - creation_date
 - rsyncFeatures
 - zfsFeatures

## Storage interactions
For storage specifically, we need:

### Source side
 - Send the requested volume through a ReadWriteCloser
 - Send the requested volume and all its snapshots through a ReadWriteCloser
 - Send the requested volume and a subset of its snapshots through a ReadWriteCloser
 - Send the requested volume snapshot through a ReadWriteCloser

As for the format of the data being sent through we effectively have 3 options:
 - rsync (fallback for filesystem)
 - dd (fallback for blocks, not currently implemented)
 - native (with current implementations for zfs, btrfs and ceph)

Which one of those is used is selected through the migration protocol.
Only native needs special handling in the driver, rsync/dd should be implemented entirely in the backend level as it's generic code on top of normal driver semantics (mounting the filesystem or mapping the block).

### Target side
On the target side, the storage backend will be provided with the instance or volume being created or refreshed, the format of the migration and a slice of snapshots to expect.

For the generic cases, if snapshots are present, a blank volume is created, the first snapshot is received and the snapshot is created, same goes for subsequent snapshots, lastly the volume itself is received and written.

For the native case, the stream is passed through to the driver along with the list of expected snapshots.

## Impact to Pool functions
Based on what the storage layer actually cares about, it feels like we should be fine with just:
```go
CreateInstanceFromMigration(i Instance, fs io.ReadWriteCloser, fsType migration.FSType, snapshots []string, op *operations.Operation) error
RefreshInstance(i Instance, fs io.ReadWriteCloser, fsType migration.FSType, snapshots []string, op *operations.Operation) error
MigrateInstance(i Instance, fs io.ReadWriteCloser, fsType migration.FSType, snapshots []string, op *operations.Operation) error

CreateCustomVolumeFromMigration(vol api.StorageVolume, fs io.ReadWriteCloser, fsType migration.FSType, snapshots []string, op *operations.Operation) error
MigrateCustomVolume(vol api.StorageVolume, fs io.ReadWriteCloser, fsType migration.FSType, snapshots []string, op *operations.Operation) error
```

So that makes things pretty nice and symmetric. The one thing that's missing though is the feature negotiation bits like we have for zfs currently. This is a driver-level thing that will need to be exposed through the backend, probably something like:

```go
MigrationFsType() (migration.MigrationFSType, []string)
```

Where the string slice is the list of supported migration features for that type.


Unless I missed something, the above should cover all our cases. The actual setup of the migration session and return of the appropriate responses (background operations with websockets) is likely better handled at the API level since we're dealing with two different uses of that code, one coming from the instance migration and the other in volume migration. So it's likely easier to have the storage package only worry about the transfer of the data itself and not have to care about the rest of the migration logic.

tomponline commented:
Added support for dir project quotas at create and copy time.

tomponline commented:
@stgraber update for today:

- Started looking at the migration code, but it seemed to be to very complex and I didn't want to make it any more complex by adding local cross-pool copy support to it.
- Instead I re-focused on adding cross-local-pool volume copying to the `lxdBackend.CreateCustomVolumeFromCopy` function.
- I've achieved the framework for doing this, and tested with cross-dir-pool copying, by adding a new struct type in the `drivers` package called `Volume`.
- The `Volume` struct achieves two things; firstly it allows me to pass all the volume info I need for both source and target volumes without having numerous arguments to the `CreateCustomVolumeFromCopy` function (and associated driver function). Secondly it allows me to embed the source pool's driver inside the `Volume` struct, allowing me to call `Volume.Mount()/Unmount()` on it, and it will then pass the request to the underlying storage driver.
- This is useful when needing to mount the source pool's volume, before doing an rsync. But it is not done in lxdBackend because it is up to the storage driver to decide whether it needs to mount to do an rsync, or whether there is a more efficient native way to do the copy. 
- I've also added a check to the existing `doVolumeCreateOrCopy` function to only use the new storage package if *both* the source pool and target pools are supported using the new package, otherwise this was causing the tests to fail due to cross-pool-copies using unsupported drivers.

Next I will continue implementing as many of the other lxdBackend and dir functions as I can, and then look to implement the same for LVM (as discussed previously). Then we can pick up the migration stuff again on your return. Thanks

lxc-jenkins commented:
Testsuite passed

lxc-jenkins commented:
Testsuite passed

lxc-jenkins commented:
Testsuite passed

lxc-jenkins commented:
Testsuite passed

lxc-jenkins commented:
Testsuite passed

tomponline commented:
@stgraber update for today:

- Added support renaming volumes and snapshots (this one took a while as I discovered a bug, and some old legacy code that lead me down the wrong path for a bit, which I've fix in a separate PR https://github.com/lxc/lxd/pull/6315 , I've also asked a question om that PR about the renaming cli help, as it hints at features that don't exist AFAIK).
- Added support for deleting snapshots.
- Added support for creating new snapshots.

I have some questions from today's work:

1. When renaming a volume, should it also rename the volume's snapshots, such that they remain under the parent volume?
2. When renaming a volume or snapshot using the dir driver, does the ext4 project quota need to have its path updated to reflect the new location?
3. Are you happy to modify `SetCustomVolumeQuota` in lxdBackend to be a more generic`UpdateCustomVolume` and `UpdateCustomVolumeSnapshot`, as this aligns more closely with how the API works today, and would allow for future additional fields without having to call a different function for each field being updated?

Additionally, I am thinking about expanding the `Volume` struct I added yesterday and moving some of the functions defined in the `Pool` interface into it. The reason for this is that in the currently defined `Pool` interface, each volume related function is separate from each other, and this will mean that any function that operates on an existing volume (such as rename, delete, rename snapshot, delete snapshot, update, update snapshot, copy, mount, unmount etc) will need to have a duplicated bit of code at the top of each function validating that the volume actually exists and loading any config needed from the database.

Instead I propose moving any of the existing volume functions from the `Pool` interface into the `Volume` struct, and adding a `GetCustomVolumeByName` function to the `Pool` interface, such that the usage would be something like this:

```go
pool, err := storagePools.GetPoolByName(d.State(), poolName)
pool.CreateCustomVolume(api.StorageVolumesPost, op)
volume := pool.GetCustomVolumeByName("myvol")
volume.Mount()
volume.Rename("mynewname")
volume.RenameSnapshot("mysnapshot", "mynewsnapshotname")
volume.Delete()
```

Is this something you would support?

tomponline commented:
@stgraber I've been spending a few hours this morning digging into the migration code to get familiar with it.

Couple of things I've noticed that I'm recording here so I don't forget:

1. Storage volume migration (via `lxc storage volume move`) always appears to use rsync and not the native send features in some storage drivers. This is via the `StorageMigrationSink` storage driver function, of which all of them just call `rsyncStorageMigrationSink`.
2. The function `rsyncStorageMigrationSink` and its sender equivalent `rsyncStorageMigrationSource` via `SendStorageVolume` do not mount the snapshot volumes either side, so for LVM migration (at least) the snapshots appear to be moved but infact are now.

tomponline commented:
@stgraber update for today, I've been focusing on migration, and have pushed a tentative implementation of the migration receiver, so that doing `lxc storage volume move...` uses the new storage driver if you're migrating to a dir storage volume. 

I'm not happy yet with where I've added the hook into the new storage layer, as it still depends on the old storage layer to create the DB row (which is a code smell in the current implementation as the volume's DB row is created in a different place to the snapshot rows strangely).

Untangling that so that the DB rows are created (or at least initiated) from the same function is my next task.

lxc-jenkins commented:
Testsuite passed

lxc-jenkins commented:
Testsuite passed

lxc-jenkins commented:
Testsuite passed

lxc-jenkins commented:
Testsuite passed

stgraber commented:
Commit title: s/implemementation/implementation/

stgraber commented on lxd/storage/drivers/driver_dir.go:
```diff
@@ -0,0 +1,864 @@
+	}
+
+	return []migration.Type{
+		{
```
Probably ought to at least be moved to something in utils.go so we can include the same migration.Type for all those using rsync.

tomponline replied:
@stgraber I'll move this function into the `common` driver, so for the drivers that use only rsync (and eventually dd for blocks) they won't need to implement this function at all.

For those that have their own native types as a preference first, they can use the following implementation:

```go
func (d *zfs) MigrationTypes(contentType ContentType) []migration.Type {
	myTypes := []migration.Type{} // My specific ZFS native type here.

        // Add the common rsync/dd types if needed here.
	myTypes = append(myTypes, d.common.MigrationTypes(contentType)...)
	return myTypes
}
```

stgraber commented on lxd/storage/drivers/driver_dir.go:
```diff
@@ -0,0 +1,864 @@
+		return fmt.Errorf("Content type not supported")
+	}
+
+	if volTargetArgs.MigrationType.FSType != migration.MigrationFSType_RSYNC {
```
We're going to have to support block eventually though, we'll effectively do that by supporting MigrationFSType_DD which will get us a simple stream of bytes from the source and then will have to pipe that through qemu-img or something to write it as qcow2 to a file on disk.

We'll need that to be able to move VMs around.

stgraber commented on lxd/storage/drivers/driver_dir.go:
```diff
@@ -0,0 +1,864 @@
+	err = rsync.Recv(path, conn, wrapper, volTargetArgs.MigrationType.Features)
+	if err != nil {
+		if ourMount {
+			vol.Unmount(op)
```
instead of having this in the normal and error path, you could just do right after the Mount call:
```go
if ourMount {
    defer vol.Unmount(op)
}
```

stgraber replied:
Actually, scrap that, we're inside the `dir` backend, we know that `MountVolume()` and `UnmountVolume()` are no-ops, so we don't even need to jump through all those hoops :)

stgraber replied:
So all the Mount and Unmount calls inside this function can go away.

tomponline replied:
@stgraber yeah agreed, my main reason for adding these unnecessary noop mount/unmount calls was to get it clear in my mind where these would need to occur in future driver types that do need mounts and that my design had accommodated for them. Happy to remove them now though.

tomponline replied:
@stgraber I thought a bit more about this, both taking into your point about removing the unnecessary mount code in dir whilst still keeping the dir implementation indicative of how other drivers should do it in the future. I've come up with adding a function to `func (v Volume) MountTask(task func(mountPath string, op *operations.Operation) error, op *operations.Operation) error` that takes a function to execute (the action you want to perform on the mounted volume) and then takes care of calling the driver's `Mount` and `Unmount` as needed (taking into account the `ourMount` response from `Mount`).

This feels like a good balance, as it reduces the mounting code in `dir`, and yet will mean that `dir` can behave similarly to other drivers that do actually need to mount.

Also, as Volume's `Mount` and `Unmount` are no longer actually used, I've removed them until they are needed.

stgraber replied:
That looks pretty clean indeed and should avoid bad uses of mount/unmount by new implementations.

stgraber commented on lxd/storage/drivers/driver_dir.go:
```diff
@@ -0,0 +1,864 @@
+
+	bwlimit := d.config["rsync.bwlimit"]
+
+	ourMount, err := srcVol.Mount(op)
```
Same comment as above, we're copying a volume within a dir pool, mount/unmount are no-ops on dir, so we can do away with all that logic.

tomponline replied:
@stgraber see comment above, have added Volume.MountTask wrapper for this.

stgraber commented:
Was thinking earlier, we need a `SupportedVolumeTypes()` function in the Driver interface.
That'll be needed so I can do `cephfs` which only supports `custom` and nothing else and have LXD return nice errors without having to special case it.

stgraber commented:
So I somehow missed that `Volume` struct earlier. Seems fine in principle, just wondering if you're actually seeing it used outside of Drivers and if so, where and why. Also, do we actually need all its attributes to be public?

stgraber commented:
Ah, I see that it is used through lxdBackend thorugh newVolume(), so indeed needs exporting from the driver package. Not sure about its fields being public still.

stgraber commented on lxd/storage/drivers/driver_dir.go:
```diff
@@ -0,0 +1,864 @@
+		return err
+	}
+
+	snapPath := GetVolumeMountPoint(d.name, volType, fmt.Sprintf("%s%s%s", volName, shared.SnapshotDelimiter, snapshotName))
```
I see this pattern going on quite a bit, should we have a GetVolumeSnapshotMountPoint function which does this?

tomponline replied:
@stgraber agreed, have added a `GetSnapshotVolumeName(parent, snapName)` function that joins them to the snapshot separator, so it can be used in other places (like db row generation) rather than just mount path generation. I'm also keen to not repeat the logic used in  GetVolumeMountPoint if possible. But this new function will make things simpler to generate a snapshot mount path with GetVolumeMountPoint.

stgraber commented on lxd/storage/drivers/driver_dir.go:
```diff
@@ -0,0 +1,864 @@
+}
+
+// RenameVolumeSnapshot renames a volume snapshot.
+func (d *dir) RenameVolumeSnapshot(volType VolumeType, volName string, newSnapshotName string, op *operations.Operation) error {
```
This now feels weird when the other snapshot functions clearly separate volName and snapName, should we just add an extra argument and have the user give us the volume, the snapshot name and the new snapshot name?

tomponline replied:
@stgraber agreed, have updated it as described.

stgraber commented on lxd/storage/backend_lxd.go:
```diff
@@ -0,0 +1,682 @@
+
+// DeleteImage removes an image from the database and underlying storage device if needed.
+func (b *lxdBackend) DeleteImage(fingerprint string, op *operations.Operation) error {
+	_, _, isSnap := shared.ContainerGetParentAndSnapshotName(fingerprint)
```
A more efficient test would probably be to confirm it's a valid sha256, that would avoid some bugs we had in the past where partial fingerprint were let through.

tomponline replied:
@stgraber have added a regex check for 64 a-z0-9 chars like we did for validating macs.

stgraber commented on lxd/storage/backend_lxd.go:
```diff
@@ -0,0 +1,682 @@
+}
+
+// CreateCustomVolumeFromMigration receives a volume being migrated.
+func (b *lxdBackend) CreateCustomVolumeFromMigration(conn *websocket.Conn, args migration.VolumeTargetArgs, op *operations.Operation) error {
```
Ah yeah, we still need to figure out a wrapper ReadWriteCloser around websocket.Conn so we can have the migration layer convert from websocket.Conn to ReadWriteCloser

stgraber commented on lxd/storage/backend_lxd.go:
```diff
@@ -0,0 +1,682 @@
+
+// RenameCustomVolume renames a custom volume and its snapshots.
+func (b *lxdBackend) RenameCustomVolume(volName string, newVolName string, op *operations.Operation) error {
+	if shared.IsSnapshot(volName) {
```
So I think it'd be good to decide who's responsibility it is to validate this stuff, whether it's on the backend (seems easier) or on the driver, but duplicating all those checks just makes the code longer with no clear benefit.

tomponline replied:
@stgraber yeah happy to do this just in backend.

stgraber commented on lxd/storage/drivers/driver_common.go:
```diff
@@ -0,0 +1,92 @@
+		"security.shifted":    shared.IsBool,
+		"security.unmapped":   shared.IsBool,
+		"volatile.idmap.last": shared.IsAny,
+		"volatile.idmap.next": shared.IsAny,
```
shifted/unmapped/volatile.idmap aren't going to be implemented at the driver level so should probably be kept away in the storage package

tomponline replied:
@stgraber I have moved this into the utils file in the storage package, but have provided it as a function that the drivers can call, so that all of the actual validation checking is kept together in the driver. The reason for this is that as part of the translation of config between drivers, sometimes fields have to be removed. This logic is kept a lot simpler if the validation and removal of unknown fields is done in a single place. It also gives the drivers the opportunity to override/remove the validation of a common field if we need to in the future. Hopefully this change achieves both the simplicity of keeping it in one place as well as avoiding defining the common fields in the drivers package.

stgraber commented on lxd/storage/utils.go:
```diff
@@ -571,9 +571,7 @@ func VolumeValidateConfig(name string, config map[string]string, parentPool *api
 
 // VolumeFillDefault fills default settings into a volume config.
 func VolumeFillDefault(name string, config map[string]string, parentPool *api.StoragePool) error {
```
This commit can probably be sent on its own?

stgraber commented on lxd/storage/drivers/volume.go:
```diff
@@ -24,3 +31,70 @@ const ContentTypeFS = ContentType("fs")
+
+// MountPoint returns the path where the volume will be mounted.
+func (v Volume) MountPoint() string {
+	return GetVolumeMountPoint(v.Pool, v.Type, v.Name)
```
Is that snapshot aware?

tomponline replied:
@stgraber yes it is.

stgraber commented on lxd/storage/drivers/volume.go:
```diff
@@ -24,3 +31,70 @@ const ContentTypeFS = ContentType("fs")
+// Snapshots returns a list of snapshots for the volume.
+func (v Volume) Snapshots(op *operations.Operation) ([]Volume, error) {
+	if v.IsSnapshot() {
+		return nil, fmt.Errorf("Volume is snapshot")
```
"is a snapshot?

stgraber commented on lxd/storage/drivers/volume.go:
```diff
@@ -24,3 +31,70 @@ const ContentTypeFS = ContentType("fs")
+// Unmount unmounts the volume.
+func (v Volume) Unmount(op *operations.Operation) (bool, error) {
+	return v.Driver.UnmountVolume(v.Type, v.Name, op)
+}
```
For Mount/Unmount, I suspect we'll need new symbols `MountVolumeSnapshot` and `UnmountVolumeSnapshot` as snapshots do need a lot of special handling for mounting/unmounting on most drivers.

We should also document the expectation that the mountpoint of a snapshot should be treated as read-only (and we may be able to make it that way by putting a read-only bind-mount on it even for dir/cephfs).

tomponline replied:
@stgraber makes sense, do you have any preference for where the "is snapshot" logic takes place: 

- In the driver's `MountVolume` logic (avoids needing new functions).
- In the `Volume` helper `Mount` function calling driver's `MountVolumeSnapshot`.
- In the driver's use of `Volume` helper, so adding `MountVolumeSnapshot` to Volume struct.

stgraber replied:
I think it's reasonable for the Volume helper to do the right thing on Mount/Unmount as it's already aware of whether it's itself a snapshot or not.

tomponline replied:
@stgraber ok cool, have added snapshot detection to volume.MountTask() and then added snapshot mount/unmount functions to driver.

stgraber commented on lxd/storage/utils.go:
```diff
@@ -486,15 +486,15 @@ var StorageVolumeConfigKeys = map[string]func(value string) ([]string, error){
 	"size": func(value string) ([]string, error) {
 		if value == "" {
-			return []string{"btrfs", "ceph", "cephfs", "lvm", "zfs"}, nil
+			return SupportedPoolTypes, nil
```
This commit can also be sent upstream already.

stgraber commented:
Overall, looks good to me. It looks like we could slim down the `dir` driver quite a bit by removing the spurious MountVolume/UnmountVolume calls and by requiring the input validation be done in the backend rather than in each driver.

We also are going to need a MountVolumeSnapshot/UnmountVolumeSnapshot of some kind.
This is something we haven't done particularly right in the past by making those mounts ephemeral but writable, leading to code which assumes that snapshots can be altered. As a result, I'd argue that even for those cases where the storage is writable, we'll want to do a read-only bind-mount during Mount and unmount that layer on Unmount so that the path is in fact read-only.

I think we can also take some good shortcuts with ZFS to access the snapshots directly rather than using the clone dance we're doing at the moment. Blocks will still need cloning if only so we can alter the filesystem uuid prior to mount, but we'll make sure to mount them read-only too so there's no confusion.

tomponline commented on lxd/storage/drivers/volume.go:
```diff
@@ -24,3 +31,70 @@ const ContentTypeFS = ContentType("fs")
 const ContentTypeBlock = ContentType("block")
+
+// Volume represents a storage volume, and provides functions to mount and unmount it.
+type Volume struct {
```
@stgraber agreed, they are now unexported.

tomponline commented:
Jenkins test this please

tomponline commented:
Jenkins please test this

tomponline commented:
jenkins: test this please

stgraber commented:
jenkins: test this please

lxc-jenkins commented:
Testsuite passed

lxc-jenkins commented:
Testsuite passed

tomponline commented:
@stgraber all checks have passed on this now.

tomponline commented:
@stgraber have rebased this on https://github.com/lxc/lxd/pull/6360
